src/coreclr/vm/jithelpers.cpp

```diff
@@ -634,72 +687,59 @@ HCIMPL1_V(UINT64, JIT_Dbl2ULngOvf, double val)
 }
 HCIMPLEND
 
-
-#if !defined(TARGET_X86) || defined(TARGET_UNIX)
-
-HCIMPL1_V(INT64, JIT_Dbl2Lng, double val)
+HCIMPL1_V(UINT32, JIT_Dbl2UInt, double val)
 {
     FCALL_CONTRACT;
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
+    const double uint_max = 4294967295.0;
+    return (val != val || val <= 0) ? 0 : (val >= uint_max) ? UINT32_MAX : (UINT32)val;
```

Comment:
Is there a way to handle both `(val != val || val <= 0)` using just one condition?

Reply:
[Clang optimizes this to one comparison already, GCC and MSVC don't though.](https://godbolt.org/z/h6Wz6oGez)

Reply:
True. Although in the C helper, we cannot combine the 2 checks ```(val != val || val <= 0)``` since any comparison with ```NaN``` will always return false.

Reply:
Why not? You just need to reverse the condition: `(val >= 0) ? .... : 0`. This should return 0 for both NaN and negative values.

Reply:
The general issue is that comparisons with NaN are not reversible, strictly speaking.

`x > NaN` returning `false` does not mean that `x <= NaN` would return `true`.

At the hardware level, `comiss` and `ucomiss` will set the flags as follows:
* `UNORDERED: ZF,PF,CF := 111;`
* `GREATER_THAN: ZF,PF,CF := 000;`
* `LESS_THAN: ZF,PF,CF := 001;`
* `EQUAL: ZF,PF,CF := 100;`

Where `UNORDERED` means one of the inputs is `NaN`. Thus, it can get optimized by a sufficiently smart compiler to one check, but reversing won't change that there are 2 checks that would be emitted by any other compiler.

Reply:
I think in this case we could make it `(val > 0) ? ((val >= uint_max) ? UINT32_MAX : (UINT32)val) : 0`

Since `val > 0` would report `false` for a `NaN` and therefore return 0, but some compilers will still emit both `jp` and `jbe` -- the JIT notably still does this too and doesn't handle all the NaN considerations "well"

Reply:
> I think in this case we could make it (val > 0) ? ((val >= uint_max) ? UINT32_MAX : (UINT32)val) : 0

Right, it is what I meant. We do this in number of other places - look for "Note that this expression also works properly for val = NaN case" comment.

Reply:
Thanks. Yes, here with ```val <= 0``` check, we can combine it and reverse the checks. Will make this change today.